server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1882,6 +1894,65 @@ public StoragePool syncStoragePool(SyncStoragePoolCmd cmd) {
         return (PrimaryDataStoreInfo) _dataStoreMgr.getDataStore(pool.getId(), DataStoreRole.Primary);
     }
 
+
+    @Override
+    public Heuristic createSecondaryStorageHeuristic(CreateSecondaryStorageSelectorCmd cmd) {
+        String name = cmd.getName();
+        String description = cmd.getDescription();
+        long zoneId = cmd.getZoneId();
+        String heuristicRule = cmd.getHeuristicRule();
+        String purpose = cmd.getPurpose();
+        HeuristicPurpose formattedPurpose = EnumUtils.getEnumIgnoreCase(HeuristicPurpose.class, purpose);
+
+        if (formattedPurpose == null) {
+            throw new IllegalArgumentException(String.format("The given heuristic purpose [%s] is not valid for creating a new secondary storage selector." +
+                    " The valid options are %s.", purpose, Arrays.asList(HeuristicPurpose.values())));
+        }
+
+        HeuristicVO heuristic = secondaryStorageHeuristicDao.findByZoneIdAndPurpose(zoneId, formattedPurpose);
```

Comment:
This is getting removed entries as well. Which means if I already created an entry then removed it, I will never be able to create another one.

Reply:
As mentioned, the method `findOneIncludingRemovedBy` only retrieves non-removed entries; therefore, this scenario does not occur.